PricesService.ts

```diff
@@ -0,0 +1,138 @@
+import { Consola } from "consola";
+import { timeout } from "promise-timeout";
+import fetchers from "./fetchers";
+import { TokensBySource } from "./ManifestParser";
+import {
+  Aggregator,
+  Credentials,
+  PriceDataAfterAggregation,
+  PriceDataBeforeAggregation,
+  PriceDataFetched
+} from "./types";
+import { trackEnd, trackStart } from "./utils/performance-tracker";
+import { v4 as uuidv4 } from 'uuid'
+
+const logger = require("./utils/logger")("PricesFetcher") as Consola;
+
+export type PricesDataFetched = { [source: string]: PriceDataFetched[] };
+export type PricesBeforeAggregation = { [token: string]: PriceDataBeforeAggregation }
+
+export default class PricesService {
+
+  constructor(
+    private fetchTimeout: number,
+    private credentials: Credentials
+  ) {
+  }
+
+  async fetchInParallel(tokensBySource: TokensBySource)
+    : Promise<PricesDataFetched[]> {
+
+    const promises: Promise<PricesDataFetched>[] = [];
+
+    for (const source in tokensBySource) {
+      promises.push(
+        this.safeFetchFromSource(
+          source,
+          tokensBySource[source]));
+    }
+
+    return await Promise.all(promises);
+  }
+
+  private async safeFetchFromSource(
+    source: string,
+    tokens: string[]): Promise<PricesDataFetched> {
+    try {
+      // Fetching
+      const pricesFromSource = await this.doFetchFromSource(source, tokens);
+
+      return {
+        source: pricesFromSource
+      }
+
+    } catch (e) {
+      // We don't throw an error because we want to continue with
+      // other fetchers even if some fetchers failed
+      const resData = e.response ? e.response.data : "";
+      logger.error(
+        `Fetching failed for source: ${source}: ${resData}`, e.stack);
+      return {};
+    }
+  }
+
+  private async doFetchFromSource(source: string, tokens: string[])
+  : Promise<PriceDataFetched[]> {
+    if (tokens.length === 0) {
+      //to później jest łapane w "safeFetch". Nie wiem, czy to dobrze
+      //bo chyba taka sytuacja świadczy o błędzie w implementacji/konfiguracji
+      //i raczej powinniśmy tutaj "fail fast"?
+      throw new Error(
+        `${source} fetcher received an empty array of symbols`);
+    }
+
+    trackStart(`fetching-${source}`);
+    const fetchPromise = fetchers[source].fetchAll(tokens, {
+      credentials: this.credentials,
+    }).then((prices) => {
+      trackEnd(`fetching-${source}`);
+      logger.info(
+        `Fetched prices in USD for ${prices.length} `
+        + `currencies from source: "${source}"`);
+      return prices;
+    });
+
+    //fail if there is no response after given timeout
```

Comment:
Not really important, but maybe we can add a single space between `//` and the comment text to keep it consistent and a bit more readable